src/Hosting.CommandLine/Internal/CliService.cs

```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using McMaster.Extensions.CommandLineUtils;
+using Microsoft.Extensions.Logging;
+
+namespace McMaster.Extensions.Hosting.CommandLine.Internal
+{
+    /// <inheritdoc/>
+    internal class CliService<T> : ICliService where T : class
+    {
+        private ILogger logger;
+        private CommandLineApplication application;
+        private CliArgs args;
+
+        /// <summary>
+        /// Creates a new instance.
+        /// </summary>
+        /// <param name="logger">A logger</param>
+        /// <param name="args">The command line arguments</param>
+        /// <param name="serviceProvider">The DI service provider</param>
+        public CliService(ILogger<CliService<T>> logger, CliArgs args,
+            IServiceProvider serviceProvider)
+        {
+            this.logger = logger;
+            this.args = args;
+
+            logger.LogDebug("Constructing CommandLineApplication<{type}> with args [{args}]",
+                typeof(T).FullName, String.Join(",", args.Value));
+            application = new CommandLineApplication<T>();
```

Comment:
Let's implement `IDisposable` on this service and dispose the application object.

Reply:
`CommandLineApplication` does not appear to be `IDisposeable`.  Am I missing something?

Reply:
`CommandLineApplication<T>` does, but not the base type....we should probably fix that too.
  https://github.com/natemcmaster/CommandLineUtils/blob/1e9bddeabed8ceb60ae60c826d6f4171042c340b/src/CommandLineUtils/CommandLineApplication%7BT%7D.cs#L15

Reply:
FYI: https://github.com/natemcmaster/CommandLineUtils/commit/aa1d1d5f5d6869c1408571ccab96f5b2f19226c8

Reply:
Again, I may be missing some C# feature, but the `Dispose` method is `private protected` thus, no access.  I notice you have this:
```
        private protected virtual void Dispose()
        {
            foreach (var command in Commands)
            {
                if (command is IDisposable dc)
                {
                    dc.Dispose();
                }
            }
        }

        void IDisposable.Dispose() => this.Dispose();
```
Which has me confused (2 methods with same signature).  So if i try:
```
        public void Dispose()
        {
            application.Dispose();
        }
```
It will fail the build:
```
Internal\CommandLineService.cs(38,25): error CS0122: 'CommandLineApplication.Dispose()' is inaccessible due to its protection level [C:\Users\lucas\git\pastdev-commandlineutils\src\Hosting.CommandLine\McMaster.Extensions.Hosting.CommandLine.csproj]
```
But I guess I can:
```
        public void Dispose()
        {
            ((IDisposable)application).Dispose();
        }
```
Was this supposed to be the [_disposable pattern_](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implementing-the-dispose-pattern-for-a-base-class) and maybe the `private protected` one is supposed to accept a `bool`?  Or did you really intend for the cast to be necessary?  Or am I missing something?

[Added the _casted call_ for now](https://github.com/natemcmaster/CommandLineUtils/pull/167/commits/554ef008b64261f116d332d3c222ea3e92072910)